### PR TITLE
Add support for flags to enhance unhardcoding mutations and bionics

### PR DESCRIPTION
**Pull request by Ramza13**
#### Summary
Infrastructure "Add support for flags to enhance unhardcoding mutations and bionics"

#### Purpose of change
Setup framework to enable easier and more powerful unhardcoding of mutations and bionics

#### Describe the solution
Add character function has_flag which checks what flags a character has on their mutations and bionics.  Trait and bionic flags already exist this is just enabling using them interchangeably.  This can be used in place of hardcoded checks for a specific mutation or bionic to allow any mutation or bionic to cause the effect.  

Additionally add active and inactive flags to both bionics and mutations.  has_flag will only return true for an active flag if the associated bionic/mutation is active and vice versa for inactive.

#### Describe alternatives you've considered
Moving it all into effects and adding ability to have bionics/mutation create permanent effects.  

#### Additional context
First commit is adding framework and converting optical dampeners.  Second commit is converting to use flag_id.
Subsequent commits are moving over more checks to use this format.

### CODE REVIEW

**Comment by actual-nh** [comment]
Can starting traits give flags?

**Comment by Ramza13** [comment]
> 
> 
> Can starting traits give flags?

Yup, and they already do.  I'll go clarify but flags already exist for mutations,/traits and bionics.  This is using them and adding new kinds.

File: src/bionics.h [hunk 1]
```diff
@@ -142,7 +142,11 @@ struct bionic_data {
     requirement_id installation_requirement;
 
     cata::flat_set<std::string> flags;
+    cata::flat_set<std::string> active_flags;
+    cata::flat_set<std::string> inactive_flags;
```
Comment (anothersimulacrum) [inline src/bionics.h:146]: Consider using `flag_id`s instead of std::strings?

Comment (Ramza13) [reply src/bionics.h:146]: Good call, I wasn't sure about those but looking closer it seems to make sense.

File: src/character.cpp [hunk 2]
```diff
@@ -392,6 +375,38 @@ static const trait_id trait_WEB_WALKER( "WEB_WALKER" );
 static const trait_id trait_WEB_WEAVER( "WEB_WEAVER" );
 
 static const std::string flag_PLOWABLE( "PLOWABLE" );
+static const flag_id json_flag_ALARMCLOCK( "ALARMCLOCK" );
```
Comment (Aivean) [inline src/character.cpp:378]: `character.cpp` already includes `flag.h`, maybe move flag constants there?

Comment (anothersimulacrum) [reply src/character.cpp:378]: We've tried static definitions in header files before (ramza specifically did, actually) - it causes a significant increase in build times.

**Comment by Aivean** [comment]
Thank you for doing this. 

One thing to note, `flag_id` (and it's `json_flag` value) are shared between too many unrelated entities. I think at some point we need to create separate `string_id` types for them to improve type safety. Obviously not in this PR.

**Review by Aivean — APPROVED** [review]

**Comment by Ramza13** [comment]
> Thank you for doing this.
> 
> One thing to note, `flag_id` (and its `json_flag` value) are shared between too many unrelated entities. I think at some point we need to create separate `string_id` types for them to improve type safety. Obviously not in this PR.

So json_flags already have a context property that seems like it's trying to be that.  I think I remember finding documentation or code saying its not actually checked at the moment, but it could be a starting point at least.  There are also documentation fields in the json so that if it's all required and used would make this cleaner.

**Comment by actual-nh** [comment]
> Thank you for doing this.

Agreed!

> One thing to note, `flag_id` (and its `json_flag` value) are shared between too many unrelated entities. I think at some point we need to create separate `string_id` types for them to improve type safety. Obviously not in this PR.

It would be nice if this could be done in a way that made searching for particular different uses of the same-named flag easier.

**Comment by Aivean** [comment]
> So json_flags already have a context property that seems like it's trying to be that.

I think "context" was originally intended to differentiate between flags applied to the different item types, as item types are represented by the same `item` class, they cannot have different types of flags.

In general, I see several classes of problems with sharing the same flag type for all entities:
1. Applying the wrong flag to entity in json
2. Checking for or setting the wrong flag to an entity in C++
3. Sharing unrelated fields in `json_flag`: (e.g. `taste_mod_`, `requires_flag_`, `inherit_`...)

Extending context could help with (1) and partially with (2), but only at runtime and at the cost of performance overhead. (3) will still be an issue and it will only grow bigger the more flag ids are used.

That's why I think using "context" is ok as a temporary solution, but ideally flags should have different type per entity. 

---

> It would be nice if this could be done in a way that made searching for particular different uses of the same-named flag easier.

I think the only way is to search for particular string and go from there. Having flags defined in one place as constants would help, but, as anothersimulacrum  pointed out, it has other drawbacks.

**Comment by Ramza13** [comment]
Ok I now changed it from using flag_id to the new json_character_flag which is just std::string.

**Comment by Aivean** [comment]
> Ok I now changed it from using flag_id to the new json_character_flag which is just std::string.

Wait, why? 

Sorry, I must have been not clear. In my opinion using any kind of `string_id` is better than using just `string` (even aliased). The eventual goal is to use separate `string_id` for flags of different entities, but I tried to empathize that this is future work.

**Comment by Ramza13** [comment]
> 
> 
> > Ok I now changed it from using flag_id to the new json_character_flag which is just std::string.
> 
> Wait, why?
> 
> Sorry, I must have been not clear. In my opinion using any kind of `string_id` is better than using just `string` (even aliased). The eventual goal is to use separate `string_id` for flags of different entities, but I tried to empathize that this is future work.

Well now making it a string_id should be a one line change.  Or would that still not be right?

**Comment by Ramza13** [comment]
Ok I was misunderstanding.  I've now changed it to be a string_id<json_flag> instead of std::string just with a different name to hopefully help keep it a little clear its a different flag.  But if I'm still not understanding just let me know and I can revert this.

**Comment by Aivean** [comment]
> Ok I was misunderstanding. I've now changed it to be a string_id<json_flag> instead of std::string just with a different name to hopefully help keep it a little clear its a different flag. But if I'm still not understanding just let me know and I can revert this.

Looks good, thank you. The only potential issue is string_id naming check in Clang-tidy. Let's see if it triggers.

**Pull request merged**